File: web/src/features/settings/connectionTone.ts
/**
 * The dot in the Settings header, from the app socket alone (`docs/DESIGN.md`
 * § "The Settings screen"): green while connected, pulsing amber while the link
 * is coming up, grey while it is down. The red the ruling gives a refused
 * gateway is the phone's: a browser whose session the gateway refuses is signed
 * out and shown the sign-in form, so it never draws this header at all.
 */
import type { DotTone } from '../../components/dotTone';
import type { SocketStatus } from '../../lib/ws';
import { strings } from '../../strings';

export function connectionTone(status: SocketStatus): DotTone {
  switch (status) {
    case 'open':
      return 'working';
    case 'connecting':
    case 'reconnecting':
      return 'waiting';
    case 'idle':
    case 'closed':
      return 'off';
  }
}

/** The word for that tone. It is read aloud and shown on hover, never printed. */
export function connectionWord(status: SocketStatus): string {
  switch (connectionTone(status)) {
    case 'working':
      return strings.settings.connected;
    case 'waiting':
      return strings.settings.connecting;
    default:
      return strings.settings.offline;
  }
}

File: ios/Sources/RCCore/State/ConnectionTone.swift
import Foundation

/// The dot beside the gateway in the Settings header, and the word for it.
///
/// `docs/DESIGN.md` § "The Settings screen": green while connected, pulsing
/// amber while the link is being made or remade, grey while it is down, red
/// when the gateway refused it. The word — Connected, Connecting, Offline,
/// Refused — is the accessibility label and nothing else: the dot carries the
/// state on screen, exactly as it does on a session row.
///
/// A pure function of the phase, so the header never decides this for itself.
public enum ConnectionTone {
    public static func dot(_ phase: ConnectionPhase) -> DotTone {
        switch phase {
        case .connected: .working
        case .connecting, .syncing, .reconnecting: .waiting
        case .signedOut: .off
        case .expired, .forbidden, .superseded, .incompatible: .failed
        }
    }

    public static func word(_ phase: ConnectionPhase) -> String {
        switch phase {
        case .connected: L10n.string("Connected")
        case .connecting, .syncing, .reconnecting: L10n.string("Connecting")
        case .signedOut: L10n.string("Offline")
        case .expired, .forbidden, .superseded, .incompatible: L10n.string("Refused")
        }
    }
}
